### DISCUSSION TITLE
Add Custom Scripts to Netbox Docker Container (And Backup)

### DISCUSSION THREAD

**Discussion by KSBolton** [netbox-community/netbox#6085] [Q&A] [answered]
Hello all. I hope this is the right place to ask this. I've just started using Netbox and I installed it the super easy way where I'm running it in a Docker container. It works great so far. Recently I wanted to include a script that would check for duplicate IPs so I hopped into bash in the container and used vi to make the .py file but the entire volume holding /etc/netbox/scripts is read-only. Is it possible to add custom scripts to the Docker version? Do I need to mess with the Dockerfile?

Separate item: I also want to backup the instance regularly. Any documented way to do that in my Docker situation? 

Any help or links to resources that could would be greatly appreciated.

**Comment by candlerb** [comment, marked as answer]
I am guessing you're running under docker-compose?  If so, and you look in the [docker-compose.yml](https://github.com/netbox-community/netbox-docker/blob/release/docker-compose.yml) file you'll see:

```
    volumes:
...
    - ./scripts:/etc/netbox/scripts:z,ro
```

So the solution is to drop the script into the `./scripts/` local directory on the host where you run docker-compose, and it will magically appear in the Netbox container.  It's read-only so that Netbox itself can't (accidentally or maliciously) modify the script.

> I also want to backup the instance regularly. Any documented way to do that in my Docker situation?

docker-compose isn't really a great way to run Netbox in production.  However, the thing you need to do is to backup the Postgres database, as described [here](https://netbox.readthedocs.io/en/stable/administration/replicating-netbox/).  You may need to tweak this to do `docker exec ...` as required.

**Reply by KSBolton** [reply]
Thank you very much. I am using docker-compose, I'll make the changes ./scripts/ today. 

What would be the best way to set up netbox in production? Would I be able to migrate what I currently have in netbox to that? Perhaps I export the Postgres DB and then reimport it into the prod environment?

**Reply by candlerb** [reply]
The recommended way of installing Netbox is given in the manual: direct install to a server or VM.  And if you want to run containers in production, then most people use kubernetes these days.

The most essential part of the stack is postgres, so if you have an existing production postgres cluster, then you can piggy-back on that.

**Reply by KSBolton** [reply]
Great, thank you again. I'll do the full install on a new host and see how best I can get the data I've imported thus far over to the new one. I don't have postgres anywhere else so I'm hoping the export and import works.

**Reply by elixdreamer** [reply]
Tried this method, but was unable to replicate this. Could this be a bug?
Since this wasn't working I have added both folders to my datastore.
`/etc/netbox/` and `/opt/netbox/netbox/scripts`
Only after adding the scripts folder to the netbox-worker i'm able to see the scripts appear in forlder `/etc/netbox/scripts`.
But these aren't appearing into the custom scripts menu.

Using the folder `/etc/netbox/scripts` as a datasource i can make it work

Please see my setup below running netboxcommunity/netbox:v4.3-3.3.0

```
  netbox:
    container_name: netbox
    volumes:
      - ./scripts:/etc/netbox/scripts:z,ro

  netbox-worker:
    volumes:
      - ./scripts:/etc/netbox/scripts:z,ro
```

**Comment by feland207** [comment]
Hello, 

I placed my script in `/opt/netbox/netbox/scripts` but I cannot see it in Netbox. Is there some special structure the Python code needs to follow? 

This is a simple script that will get list of devices + IP interfaces from Netbox and will generate a config template.

**Reply by candlerb** [reply]
You're replying to an existing thread about scripts in docker containers.  Are you using docker?  If not, I think the question would be better in a new thread.

But in short: yes, in a standalone Netbox installation where you haven't modified `SCRIPTS_ROOT`, you just drop your script in that directory.  It needs to follow the correct structure for a custom script, which is to define a class which is a subclass of `extras.scripts.Script`.

Try using one of the examples, e.g. https://github.com/netbox-community/reports/blob/master/scripts/create_vm.py

**Reply by feland207** [reply]
Yes, I am using docker containers. 

I took your example from [here](https://github.com/netbox-community/reports/blob/master/scripts/create_vm.py), hop into my container `netbox-docker_netbox_1` and add `.py` file under `/opt/netbox/netbox/scripts` 

Then I browse to Netbox page in `Other/Scripts` and it does not appear there magically.  Is any extra needed or logs to check to troubleshoot this?

**Reply by candlerb** [reply]
It's explained in the [previous posting](https://github.com/netbox-community/netbox/discussions/6085#discussioncomment-565017):  the docker container expects scripts to be in `/etc/netbox/scripts/`.  This is set in the custom [configuration.py](https://github.com/netbox-community/netbox-docker/blob/release/configuration/configuration.py#L224-L226) that the docker container comes with.
